src/wasm/wat-parser.cpp

```diff
@@ -813,6 +1065,602 @@ Result<typename Ctx::GlobalTypeT> globaltype(Ctx& ctx, ParseInput& in) {
   RETURN_OR_OK((GlobalType{Immutable, *type}));
 }
 
+// ============
+// Instructions
+// ============
+
+template<typename Ctx>
+MaybeResult<typename Ctx::InstrT> instr(Ctx& ctx, ParseInput& in) {
+  auto keyword = in.takeKeyword();
+  if (!keyword) {
+    return {};
+  }
+
+  auto op = *keyword;
+
+#define NEW_INSTRUCTION_PARSER
+#define NEW_WAT_PARSER
+#include <gen-s-parser.inc>
+}
+
+template<typename Ctx>
+Result<typename Ctx::InstrsT> instrs(Ctx& ctx, ParseInput& in) {
+  // TODO: Folded instructions.
+  std::vector<Expression*> insts;
+  while (auto inst = instr(ctx, in)) {
+    CHECK_ERR(inst);
+    if constexpr (parsingDefs<Ctx>) {
+      insts.push_back(*inst);
+    }
+  }
+  if constexpr (parsingDefs<Ctx>) {
+    return insts;
+  } else {
+    return Ok{};
+  }
+}
+
+template<typename Ctx>
+Result<typename Ctx::ExprT> expr(Ctx& ctx, ParseInput& in) {
+  auto insts = instrs(ctx, in);
+  CHECK_ERR(insts);
+  if constexpr (parsingDefs<Ctx>) {
+    switch (insts->size()) {
+      case 0:
+        return ctx.builder.makeNop();
+      case 1:
+        return insts->front();
+      default:
+        return ctx.builder.makeBlock(*insts);
+    }
+  } else {
+    return Ok{};
+  }
+}
+
+template<typename Ctx> Result<typename Ctx::InstrT> makeUnreachable(Ctx& ctx) {
+  if constexpr (parsingDefs<Ctx>) {
+    return ctx.builder.makeUnreachable();
+  } else {
+    return Ok{};
+  }
+}
+
+template<typename Ctx> Result<typename Ctx::InstrT> makeNop(Ctx& ctx) {
+  if constexpr (parsingDefs<Ctx>) {
+    return ctx.builder.makeNop();
+  } else {
+    return Ok{};
+  }
+}
```

Comment:
Does SFINAE or such let us write a template def more than once? if so we could emit a failing stub for all the instructions (using `wasm-delegations.def`) and then just add the working ones after that.